src/scheduler/fifo.c

```diff
@@ -1094,18 +1039,40 @@ main_sched_loop(status *policy, int sd, server_info *sinfo, schd_error **rerr)
 		}
 
 		if (!end_cycle) {
-			if (second_connection != -1) {
-				char *jid = NULL;
-
-				/* get_sched_cmd_noblk() located in file get_4byte.c */
-				if ((get_sched_cmd_noblk(second_connection, &cmd, &jid) == 1) &&
-					(cmd == SCH_SCHEDULE_RESTART_CYCLE)) {
-					log_event(PBSEVENT_SCHED, PBS_EVENTCLASS_JOB, LOG_WARNING,
-						njob->name, "Leaving scheduling cycle as requested by server.");
-					end_cycle = 1;
+			if (servers) {
+				int i = 0;
+				for (; servers[i] != NULL; i++) {
+					sched_cmd *cmd;
+					if (servers[i]->secondary_sock <= 0)
+						continue;
+					cmd = malloc(sizeof(sched_cmd));
+					if (cmd == NULL) {
+						log_err(errno, __func__, MEM_ERR_MSG);
+						continue;
+					}
+					cmd->jid = NULL;
+					if (get_sched_cmd_noblk(servers[i]->secondary_sock, cmd) != 1) {
```

Comment:
Here also I think we need to handle the case of server going down i.e. if get_sched_cmd_noblok returns EOF.

Reply:
I am not sure what to do here if we detect EOF as its main_sched_loop() and I still understanding internals of sched. May be you or @bhroam can guide me what should we do here if we detect EOF from server side here?

Reply:
Just my thought. I think it is better to end the current cycle when we get EOF here which indicates server is down. Better to catch server going down early. Even if it is ignored here it can fail further for example like while sending attribute updates or when it calls any IFL call after this because it can get SIGPIPE. And also our final goal is we should not continue scheduling even when one server is down. 

Earlier case is little different because super high priority commands were only sent on secondary connection, but now we send both normal/medium/super high priority commands on the same secondary connection.

Having said that,  Let us take Bhroam's suggestion on this and based on his inputs we can proceed accordingly. @bhroam can you please let us know your opinion.

Reply:
Usually what happens when the server goes down is we try and write over the primary connection and get a SIGPIPE.  We take note of this and exit the cycle.  Now that we are in charge of opening the connections, just exiting the cycle won't be enough.  Won't we need to reopen those connections?  In any case, you will need to return a value from this function that we can catch in main_sched_loop and set end_cycle = 1.

Reply:
Also note that there is issue in get_sched_cmd_noblk() then it doesn't differenciate between 'no super high prio command from server' vs EOF on closed connection. In both case it returns 0 only.
@bhroam (Assuming we are able to fix get_sched_cmd_noblk()), How about we do something like in main_sched_loop():
```c
if (!end_cycle) {
	int rc = -1;
	sched_cmd *cmd = get_high_prio_cmd(&rc);
	if (rc == 0) {
		log_event(PBSEVENT_SCHED, PBS_EVENTCLASS_JOB, LOG_WARNING,
				njob->name, "Looks like we lost connection with one of the server, leaving scheduling cycle");
		end_cycle = 1;
	} else if (cmd && cmd->cmd == SCH_SCHEDULE_RESTART_CYCLE) {
		log_event(PBSEVENT_SCHED, PBS_EVENTCLASS_JOB, LOG_WARNING,
				njob->name, "Leaving scheduling cycle as requested by server.");
		end_cycle = 1;
	}
	free_sched_cmd(cmd);
}
```

Reply:
This looks fine to me.  I'd rephrase the first message slightly to be more definite.  Drop the 'looks like' part.